String Operation Split() in WHERE

Please help to find a proper way to use Split() in Where clause. I am trying to
`.Where(n => n.OrganizationName.Value.Split(/* :( */).Any(t => t.ToLower().StartsWith(match)))`
All documented and undocumented overloads are to no avail: System.NotSupportedException: Could not parse expression 'n.OrganizationName.Value.Split(System.Char[])': This overload of the method 'System.String.Split' is currently not supported.

Thank you
P.S: I expect Split() translated to TOKENS N1QL function.

Passing a single space character
`.Where(n => n.OrganizationName.Value.Split(' ').Any(t => t.ToLower().StartsWith(match)))`
 I am getting build error CS0854: An expression tree may not contain a call or invocation that uses optional arguments.
Setting the optional argument explicitly
` .Where(n => n.OrganizationName.Value.Split(' ', StringSplitOptions.None).Any(t => t.ToLower().StartsWith(match)))`
I am getting a runtime error: System.NotSupportedException: Could not parse expression. :(

--- a/Src/Couchbase.Linq.UnitTests/QueryGeneration/ArrayOperatorTests.cs
+++ b/Src/Couchbase.Linq.UnitTests/QueryGeneration/ArrayOperatorTests.cs
@@ -1,12 +1,9 @@
-﻿using System;
-using System.Collections.Generic;
+﻿using System.Collections.Generic;
 using System.Linq;
-using Couchbase.Core;
 using Couchbase.Linq.UnitTests.Documents;
 using Moq;
 using Newtonsoft.Json.Serialization;
 using NUnit.Framework;
-using Couchbase.Linq.Extensions;
 
 namespace Couchbase.Linq.UnitTests.QueryGeneration
 {
@@ -105,6 +102,63 @@ namespace Couchbase.Linq.UnitTests.QueryGeneration
                     .Where(e => staticArray.Contains(e.Name));
 
             const string expected = "SELECT RAW `Extent1` FROM `default` as `Extent1` WHERE `Extent1`.`name` IN (['abc', 'def'])";
+
+            var n1QlQuery = CreateN1QlQuery(mockBucket.Object, query.Expression);
+
+            Assert.AreEqual(expected, n1QlQuery);
+        }
+
+        [Test]
+        public void Test_ArrayAny()
+        {
+            SetContractResolver(new DefaultContractResolver());
+
+            var mockBucket = new Mock<IBucket>();
+            mockBucket.SetupGet(e => e.Name).Returns("default");
+
+            var query =
+                QueryFactory.Queryable<Brewery>(mockBucket.Object)
+                    .Where(e => e.Beers.Any(p => p == "test"));
+
+            const string expected = "SELECT RAW `Extent1` FROM `default` as `Extent1` WHERE ANY `Extent2` IN `Extent1`.`beers` SATISFIES (`Extent2` = 'test') END";
+
+            var n1QlQuery = CreateN1QlQuery(mockBucket.Object, query.Expression);
+
+            Assert.AreEqual(expected, n1QlQuery);
+        }
+
+        [Test]
+        public void Test_StringSplitAny()
+        {
+            SetContractResolver(new DefaultContractResolver());
+
+            var mockBucket = new Mock<IBucket>();
+            mockBucket.SetupGet(e => e.Name).Returns("default");
+
+            var query =
+                QueryFactory.Queryable<Beer>(mockBucket.Object)
+                    .Where(e => e.Name.Split().Any(p => p == "test"));
+
+            const string expected = "SELECT RAW `Extent1` FROM `default` as `Extent1` WHERE ANY `Extent2` IN SPLIT(`Extent1`.`name`) SATISFIES (`Extent2` = 'test') END";
+
+            var n1QlQuery = CreateN1QlQuery(mockBucket.Object, query.Expression);
+
+            Assert.AreEqual(expected, n1QlQuery);
+        }
+
+        [Test]
+        public void Test_StringSplitTwoParamAny()
+        {
+            SetContractResolver(new DefaultContractResolver());
+
+            var mockBucket = new Mock<IBucket>();
+            mockBucket.SetupGet(e => e.Name).Returns("default");
+
+            var query =
+                QueryFactory.Queryable<Beer>(mockBucket.Object)
+                    .Where(e => e.Name.Split(new[] {' '}).Any(p => p == "test"));
+
+            const string expected = "SELECT RAW `Extent1` FROM `default` as `Extent1` WHERE ANY `Extent2` IN SPLIT(`Extent1`.`name`, ' ') SATISFIES (`Extent2` = 'test') END";
 
             var n1QlQuery = CreateN1QlQuery(mockBucket.Object, query.Expression);
 

--- a/Src/Couchbase.Linq/Clauses/ArrayGeneratingFunctionExpressionNode.cs
+++ b/Src/Couchbase.Linq/Clauses/ArrayGeneratingFunctionExpressionNode.cs
@@ -1,0 +1,79 @@
+﻿using System;
+using System.Linq.Expressions;
+using Remotion.Linq;
+using Remotion.Linq.Clauses;
+using Remotion.Linq.Clauses.Expressions;
+using Remotion.Linq.Parsing.Structure.IntermediateModel;
+using Remotion.Linq.Utilities;
+
+namespace Couchbase.Linq.Clauses
+{
+    internal sealed class ArrayGeneratingFunctionExpressionNode : IQuerySourceExpressionNode
+    {
+        private readonly Type _querySourceElementType;
+        public Type QuerySourceElementType => _querySourceElementType;
+
+        public Type QuerySourceType { get; }
+        public Expression ParsedExpression { get; }
+        public string AssociatedIdentifier { get; }
+        public IExpressionNode Source => null;
+
+        public ArrayGeneratingFunctionExpressionNode(MethodCallExpressionParseInfo parseInfo, Expression argument1)
+        {
+            QuerySourceType = parseInfo.ParsedExpression.Type;
+
+            if (!ItemTypeReflectionUtility.TryGetItemTypeOfClosedGenericIEnumerable(parseInfo.ParsedExpression.Type,
+                out _querySourceElementType))
+            {
+                _querySourceElementType = typeof(object);
+            }
+
+            ParsedExpression = parseInfo.ParsedExpression;
+            AssociatedIdentifier = parseInfo.AssociatedIdentifier;
+        }
+
+        public Expression Resolve(ParameterExpression inputParameter, Expression expressionToBeResolved,
+            ClauseGenerationContext clauseGenerationContext)
+        {
+            if (inputParameter == null)
+            {
+                throw new ArgumentNullException(nameof(inputParameter));
+            }
+            if (expressionToBeResolved == null)
+            {
+                throw new ArgumentNullException(nameof(expressionToBeResolved));
+            }
+
+            return QuerySourceExpressionNodeUtility.ReplaceParameterWithReference(
+                this,
+                inputParameter,
+                expressionToBeResolved,
+                clauseGenerationContext);
+        }
+
+        public QueryModel Apply(QueryModel queryModel, ClauseGenerationContext clauseGenerationContext)
+        {
+            if (queryModel != null)
+            {
+                throw new ArgumentException(
+                    "QueryModel has to be null because MainSourceExpressionNode marks the start of a query.",
+                    nameof(queryModel));
+            }
+
+            var mainFromClause = CreateMainFromClause(clauseGenerationContext);
+            var defaultSelectClause = new SelectClause(new QuerySourceReferenceExpression(mainFromClause));
+            return new QueryModel (mainFromClause, defaultSelectClause) { ResultTypeOverride = QuerySourceType };
+        }
+
+        private MainFromClause CreateMainFromClause(ClauseGenerationContext clauseGenerationContext)
+        {
+            var fromClause = new MainFromClause (
+                AssociatedIdentifier,
+                QuerySourceElementType,
+                ParsedExpression);
+
+            clauseGenerationContext.AddContextInfo(this, fromClause);
+            return fromClause;
+        }
+    }
+}

--- a/Src/Couchbase.Linq/QueryGeneration/MethodCallTranslators/StringSplitMethodCallTranslator.cs
+++ b/Src/Couchbase.Linq/QueryGeneration/MethodCallTranslators/StringSplitMethodCallTranslator.cs
@@ -10,7 +10,7 @@ namespace Couchbase.Linq.QueryGeneration.MethodCallTranslators
 {
     internal class StringSplitMethodCallTranslator : IMethodCallTranslator
     {
-        private static readonly MethodInfo[] SupportedMethodsStatic =
+        public static readonly MethodInfo[] SupportedMethodsStatic =
         {
             typeof (string).GetMethod("Split", new[] { typeof (char[]) })
         };

--- a/Src/Couchbase.Linq/QueryGeneration/N1QLQueryModelVisitor.cs
+++ b/Src/Couchbase.Linq/QueryGeneration/N1QLQueryModelVisitor.cs
@@ -171,9 +171,9 @@ namespace Couchbase.Linq.QueryGeneration
                 // This is an Array type subquery, since we're querying against a member not a bucket
                 _queryPartsAggregator.QueryType = N1QlQueryType.Array;
             }
-            else if (fromClause.FromExpression is SubQueryExpression)
+            else if (fromClause.FromExpression is SubQueryExpression subQueryExpression)
             {
-                VisitSubQueryFromClause(fromClause, (SubQueryExpression) fromClause.FromExpression);
+                VisitSubQueryFromClause(fromClause, subQueryExpression);
             }
             else if (fromClause.FromExpression is QuerySourceReferenceExpression querySourceReferenceExpression)
             {
@@ -196,9 +196,9 @@ namespace Couchbase.Linq.QueryGeneration
                     throw new NotSupportedException("From Clause Is Referencing An Invalid Query Source");
                 }
             }
-            else if (fromClause.FromExpression is ConstantExpression)
+            else if (fromClause.FromExpression is ConstantExpression || fromClause.FromExpression is MethodCallExpression)
             {
-                // From clause for this subquery is a constant array
+                // From clause for this subquery is a constant array or a function returning an array
 
                 VisitArrayFromClause(fromClause);
             }

--- a/Src/Couchbase.Linq/QueryParserHelper.cs
+++ b/Src/Couchbase.Linq/QueryParserHelper.cs
@@ -1,7 +1,10 @@
-﻿using Couchbase.Linq.Clauses;
+﻿using System.Linq;
+using Couchbase.Linq.Clauses;
 using Couchbase.Linq.Extensions;
 using Couchbase.Linq.Operators;
+using Couchbase.Linq.QueryGeneration;
 using Couchbase.Linq.QueryGeneration.ExpressionTransformers;
+using Couchbase.Linq.QueryGeneration.MethodCallTranslators;
 using Couchbase.Linq.Serialization;
 using Couchbase.Linq.Utils;
 using Remotion.Linq.Parsing.ExpressionVisitors.Transformation;
@@ -50,6 +53,9 @@ namespace Couchbase.Linq
             nodeTypeRegistry.Register(AverageAsyncExpressionNode.GetSupportedMethods(), typeof(AverageAsyncExpressionNode));
             nodeTypeRegistry.Register(MinAsyncExpressionNode.GetSupportedMethods(), typeof(MinAsyncExpressionNode));
             nodeTypeRegistry.Register(MaxAsyncExpressionNode.GetSupportedMethods(), typeof(MaxAsyncExpressionNode));
+
+            // register ArrayGeneratingFunctionExpressionNode
+            nodeTypeRegistry.Register(StringSplitMethodCallTranslator.SupportedMethodsStatic, typeof(ArrayGeneratingFunctionExpressionNode));
 
             //This creates all the default node types
             var nodeTypeProvider = ExpressionTreeParser.CreateDefaultNodeTypeProvider();

